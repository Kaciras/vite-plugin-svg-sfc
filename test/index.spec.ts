import { cwd } from "process";
import { basename, join } from "path";
import { readFileSync, writeFileSync } from "fs";
import { expect, it } from "vitest";
import { build, createServer } from "vite";
import { RollupOutput } from "rollup";
import vue from "@vitejs/plugin-vue";
import { createApp } from "vue";
import { renderToString } from "vue/server-renderer";
import {
	compile,
	convert,
	copyFixture,
	resolveFixture,
	TestOptions,
	useTempDirectory,
	ViteHMRClient,
} from "./test-utils";
import svgSfc from "../index.js";

const input = "image.svg";
const tmpDir = useTempDirectory(cwd());

const strokeSVG = readFileSync(resolveFixture("stroke.svg"), "utf8");

async function loadBundle<T = any>(code: string) {
	const file = join(tmpDir, "test.js");
	writeFileSync(file, code);
	return (await import("file://" + file)).default as T;
}

it("should throw on non-SVG data", async () => {
	const build = convert("png-data.svg?sfc");
	await expect(build).rejects.toThrow();
});

it("should throw on non-exists file", async () => {
	const build = convert("not exists.svg?sfc");
	await expect(build).rejects.toThrow();
});

it("should keep query in the URL", async () => {
	const bundle = await compile("stroke.svg?foo=1&sfc&bar");
	const absPath = resolveFixture("stroke.svg").replaceAll("\\", "/");
	expect(bundle.output[0].facadeModuleId)
		.toBe(absPath + ".vue?foo=1&sfc&bar");
});

it("should change attributes", async () => {
	expect(await convert("styles-0.svg?sfc")).toMatchSnapshot();
});

it("should change stroke attribute", async () => {
	expect(await convert("stroke.svg?sfc")).toMatchSnapshot();
});

it("should remove processing instruction in dev", async () => {
	expect(await convert("instruction.svg?sfc", { mode: "development" })).toMatchSnapshot();
});

it("should remove processing instruction in prod", async () => {
	expect(await convert("instruction.svg?sfc")).toMatchSnapshot();
});

it("should not minify on dev mode", async () => {
	expect(await convert("stroke.svg?sfc", { mode: "development" })).toMatchSnapshot();
});

it("should support set minify in options", async () => {
	const code = await convert("stroke.svg?sfc", {
		mode: "production",
		config: { minify: false },
	});
	expect(code).toMatch("<title>Untitled-1</title>");
});

it("should support custom mark", async () => {
	const code = await convert("styles-0.svg?component", { 
		config: { mark: "component" },
	});
	expect(code).toMatchSnapshot();
});

it("should extract styles", async () => {
	const code = await convert("styles-0.svg?sfc");
	expect(code.toString()).toMatchSnapshot();
});

it("should not process SVG if svgo option is false", async () => {
	const config = { svgo: false } as const;
	const code = await convert("stroke.svg?sfc", { config });
	expect(code).toBe(`<template>${strokeSVG}</template>`);
});

it("should work with @vitejs/plugin-vue", async () => {
	const bundle = await build({
		logLevel: "silent",
		build: {
			write: false,
			ssr: resolveFixture("styles-0.svg?sfc"),
		},
		plugins: [vue(), svgSfc()],
	});
	const { code } = (bundle as RollupOutput).output[0];

	const component = await loadBundle(code);
	const app = createApp(component, { width: 4396 });
	expect(await renderToString(app)).toMatchSnapshot();
});

it("should apply extractCSS plugin", async () => {
	const code = await convert("styles-0.svg?sfc", {
		config: {
			svgo: { plugins: ["extractCSS"] },
		},
	});
	expect(code).toMatchSnapshot();
});

it("should change <svg>'s attributes with svgProps", async () => {
	const config: TestOptions = {
		config: {
			svgProps: {
				":data-foo": "1",	// Add new
				viewBox: "0 0 5 5",	// Replace
			},
		},
	};
	expect(await convert("styles-0.svg?sfc", config)).toMatchSnapshot();
});

it("should change <svg>'s attributes with custom function", async () => {
	const config: TestOptions = {
		config: {
			svgProps(attrs, path, passes) {
				attrs.path = basename(path);
				attrs.passes = passes;
			},
		},
	};
	expect(await convert("styles-0.svg?sfc", config)).toMatchSnapshot();
});

it("should support HMR", async () => {
	const filename = join(tmpDir, "image.svg");
	const mainUrl = `/${input}?sfc`;
	const styleUrl = `/${filename}.vue?vue&type=style&index=0&scoped=true&lang.css`;

	copyFixture("styles-0.svg", filename);

	const server = await createServer({
		logLevel: "silent",
		root: tmpDir,
		build: {
			rollupOptions: {
				input: input + "?sfc",
			},
		},
		plugins: [svgSfc(), vue()],
	});

	async function getStyleCode() {
		await server.transformRequest(mainUrl);
		return (await server.transformRequest(styleUrl))?.code;
	}

	await server.listen();
	const client = new ViteHMRClient(server);

	expect(await getStyleCode()).contains("fill: blue;");

	const waitForHMR = client.receive();
	copyFixture("styles-1.svg", filename);
	const { updates } = await waitForHMR;

	const style = updates.find(e => e.path === styleUrl);
	expect(style?.type).toBe("js-update");
	expect(await getStyleCode()).contains("fill: red;");

	await server.close();
});
