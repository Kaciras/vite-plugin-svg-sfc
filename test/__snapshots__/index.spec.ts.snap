// Vitest Snapshot v1

exports[`should apply extractCSS plugin 1`] = `"<template><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 20 20\\"><defs/><rect id=\\"rect\\" width=\\"10\\" height=\\"10\\"/></svg></template><style scoped>#rect { fill: blue; }.st0 { width: 100px; }</style>"`;

exports[`should change <svg>'s attributes with svgProps 1`] = `"<template><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 5 5\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" :data-foo=\\"1\\"><path id=\\"rect\\" d=\\"M0 0h10v10H0z\\"/></svg></template><style scoped>#rect{fill:#00f}.st0{width:100px}</style>"`;

exports[`should change attributes 1`] = `"<template><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 20 20\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\"><path id=\\"rect\\" d=\\"M0 0h10v10H0z\\"/></svg></template><style scoped>#rect{fill:#00f}.st0{width:100px}</style>"`;

exports[`should change stroke attribute 1`] = `"<template><svg id=\\"layer_1\\" data-name=\\"layer 1\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 92.01 92.55\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\"><path d=\\"m90.02 66.22-31.11-.97-18.43 25.08-8.69-29.89-29.55-9.78 25.74-17.5.17-31.13 24.6 19.07 29.65-9.45-10.54 29.28 18.16 25.29z\\"/></svg></template>"`;

exports[`should extract styles 1`] = `"<template><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 20 20\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\"><path id=\\"rect\\" d=\\"M0 0h10v10H0z\\"/></svg></template><style scoped>#rect{fill:#00f}.st0{width:100px}</style>"`;

exports[`should not minify on dev mode 1`] = `"<template><svg id=\\"layer_1\\" data-name=\\"layer 1\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 92.01 92.55\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\"><title>Untitled-1</title><polygon points=\\"90.02 66.22 58.91 65.25 40.48 90.33 31.79 60.44 2.24 50.66 27.98 33.16 28.15 2.03 52.75 21.1 82.4 11.65 71.86 40.93 90.02 66.22\\"/></svg></template>"`;

exports[`should remove processing instruction in dev 1`] = `"<template><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\"><path d=\\"M0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0z\\"/></svg></template>"`;

exports[`should remove processing instruction in prod 1`] = `"<template><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\"><path d=\\"M0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0z\\"/></svg></template>"`;

exports[`should work with @vitejs/plugin-vue 1`] = `"<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewbox=\\"0 0 20 20\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"4396\\" data-v-61680a2e><path id=\\"rect\\" d=\\"M0 0h10v10H0z\\" data-v-61680a2e></path></svg>"`;
